api/src/sockets/kots/DeploySocket.ts

```diff
@@ -114,99 +115,137 @@ export class KotsDeploySocketService {
+    if (!app.restoreInProgressName) {
+      return;
+    }
 
-              this.io.in(clusterSocketHistory.clusterId).emit("deploy", args);
+    const velero = new VeleroClient("velero"); // TODO velero namespace
 
-              // reset app deployment state
-              clusterSocketHistory.sentDeploySequences = _.filter(clusterSocketHistory.sentDeploySequences, (s) => {
-                return !_.startsWith(s, app.id);
-              });
+    try {
+      const restore = await velero.readRestore(app.restoreInProgressName);
+      if (!restore) {
+        await this.startVeleroRestore(velero, app);
+      } else {
+        await this.checkRestoreComplete(velero, restore, app);
+      }
+    } catch (err) {
+      logger.warn("Velero restore failed");
+      logger.warn(err);
+    }
+  }
 
-              await this.kotsAppStore.updateAppRestoreUndeployStatus(app.id, "in_process");
-            } catch (err) {
-              console.log("Restore undeploy failed");
-              console.log(err);
-            }
-            break;
-          }
+  async startVeleroRestore(velero: VeleroClient, app: KotsApp): Promise<void> {
+    if (!app.restoreInProgressName) {
+      return;
+    }
+
+    logger.info(`Creating velero Restore object ${app.restoreInProgressName}`);
+
+    // create the Restore resource
+    const snapshotName = getSnapshotNameFromRestoreName(app.restoreInProgressName);
+    await velero.restore(snapshotName, app.restoreInProgressName);
+  }
+
+  async checkRestoreComplete(velero: VeleroClient, restore: Restore, app: KotsApp) {
+    switch (_.get(restore, "status.phase")) {
+      case Phase.Completed:
+        // Switch operator back to deploy mode on the restored sequence
+        const backup = await velero.readBackup(restore.spec.backupName);
+        if (!backup.metadata.annotations) {
+          throw new ReplicatedError(`Backup is missing required annotations`);
         }
-      }
+        const sequenceString = backup.metadata.annotations[kotsAppSequenceKey];
+        if (!sequenceString) {
+          throw new ReplicatedError(`Backup is missing sequence annotation`);
+        }
+        const sequence = parseInt(sequenceString, 10);
+        if (_.isNaN(sequence)) {
+          throw new ReplicatedError(`Failed to parse sequence from Backup: ${sequenceString}`);
+        }
+        const clusterId = backup.metadata.annotations[kotsClusterIdKey];
+        if (!clusterId) {
+          throw new ReplicatedError(`Backup is missing cluster ID annotation`);
+        }
+
+        logger.info(`Restore complete, setting deploy version to ${sequence}`);
+        await this.kotsAppStore.deployVersion(app.id, sequence, clusterId);
+        await this.kotsAppStore.updateAppRestoreReset(app.id);
+        break;
+
+      case Phase.PartiallyFailed:
+      case Phase.Failed:
+        logger.info(`Restore failed, resetting app restore name`);
+        await this.kotsAppStore.updateAppRestoreReset(app.id);
```

Comment:
We have a pending task `Add a cancel restore button/mutation with warning`. I think a user should have to ack the failed restore with that mutation before it's reset, otherwise they're likely to think it succeeded.

Reply:
Can we treat this as a separate issue?

Reply:
yes